website/blog/2022-04-14-add-ci-cd-to-bitbucket.md

```diff
@@ -0,0 +1,289 @@
+3. Bitbucket (or other platform) environment - configure our environment with the necessary secrets and variables to run our workflow 
+
+## Step 1: Database preparation
+
+In general, we want the following:
+
+1. Create a CI user with proper grants in your database, including the ability to create the schema(s) they’ll write to (`create on [dbname]`).
+2. Create a production user with proper grants in your database.
+
+The specifics will differ based on the database type you’re using. To see what constitutes “proper grants”, please consult the dbt Discourse classic “[The exact grant statements we use in a dbt project](https://discourse.getdbt.com/t/the-exact-grant-statements-we-use-in-a-dbt-project/430)” and Matt Mazur’s “[Wrangling dbt Database Permissions](https://mattmazur.com/2018/06/12/wrangling-dbt-database-permissions/)”.
+
+In my case, I created:
+
+1. A `dev_ci` Postgres user which had been granted a previously created `role_dev` role (same as all other development users). `role_dev` has connect and create grants on the database.
+2. A `dbt_bitbucket` user which has been granted a previously created `role_prod` role (same as dbt Cloud production environment). The `role_prod` role must have write access to your production schemas.
+
+```sql
+create role role_dev;
+
+grant create on database [dbname] to role_dev;
+
+-- Grant all permissions required for the development role
+
+create role role_prod;
+
+grant create on database [dbname] to role_prod;
+
+-- Grant all permissions required for the production role
+
+create role dev_ci with login password ‘[password]’;
+
+grant role_dev to dev_ci;
+
+create schema dbt_ci;
+
+grant all on schema dbt_ci to role_dev;
+
+alter schema dbt_ci owner to role_dev;
+
+create role dbt_bitbucket with login password ‘[password]’;
+
+grant role_prod to dbt_bitbucket;
+```
+
+### Role Masking
+
+Finally - and this might be a Postgres-only step - I had to make sure that the regularly scheduled dbt Cloud jobs connected with a `dbt_cloud` user with a `role_prod` grant would be able to drop and re-create views and tables during their run, which they could not if `dbt_bitbucket` had previously created and owned them. To do that, I needed to [mask both roles](https://dba.stackexchange.com/a/295736):
+
+```sql
+alter role dbt_bitbucket set role role_prod;
+
+alter role dbt_cloud set role role_prod;
+```
+
+That way, any tables and views created by either user would be owned by “user” role_prod.
+
+## Step 2: Repository preparation
+
+Next, we’ll need to configure the repository. Within the repo, we’ll need to configure:
+
+1. The pipeline environment
+2. The database connections
+3. The pipeline itself
+
+### Pipeline environment: requirements.txt
+
+You’ll need at least your dbt adapter-specific package, ideally pinned to a version. Mine is just
+
+`dbt-[adapter] ~= 1.0`.
```

Comment:
Should this be a code block?

Reply:
ehhh yea, yea, for consistency, even though it's one line it's still an entire file and we should make it a block.